File: app/src/main/java/org/ole/planet/myplanet/model/RealmApkLog.java
package org.ole.planet.myplanet.model;

import android.content.Context;

import com.google.gson.JsonObject;

import org.ole.planet.myplanet.utilities.NetworkUtils;

import io.realm.RealmObject;
import io.realm.annotations.Ignore;
import io.realm.annotations.PrimaryKey;

public class RealmApkLog extends RealmObject {
    @Ignore
    public static final String ERROR_TYPE_EXCEPTION = "exception";
    @Ignore
    public static final String ERROR_TYPE_CRASH = "crash";
    @Ignore
    public static final String ERROR_TYPE_ANR = "AnR";
    @PrimaryKey
    private String id;
    private String type;
    private String _rev;
    private String error;
    private String page;
    private String parentCode;
    private String version;
    private String createdOn;
    private String time;
    public String getId() {
        return id;
    }

    public String getParentCode() {
        return parentCode;
    }

    public void setParentCode(String parentCode) {
        this.parentCode = parentCode;
    }

    public String getCreatedOn() {
        return createdOn;
    }

    public void setCreatedOn(String createdOn) {
        this.createdOn = createdOn;
    }


    public String getTime() {
        return time;
    }

    public void setTime(String time) {
        this.time = time;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String get_rev() {
        return _rev;
    }

    public void set_rev(String _rev) {
        this._rev = _rev;
    }

    public String getType() {
        return type;
    }

    public void setType(String type) {
        this.type = type;
    }

    public String getError() {
        return error;
    }

    public String getVersion() {
        return version;
    }

    public void setVersion(String version) {
        this.version = version;
    }

    public static JsonObject serialize(RealmApkLog log, Context context) {
        JsonObject object = new JsonObject();
        object.addProperty("type", log.getType());
        object.addProperty("error", log.getError());
        object.addProperty("page", log.getPage());
        object.addProperty("time", log.getTime());
        object.addProperty("version", log.getVersion());
        object.addProperty("createdOn", log.getCreatedOn());
        object.addProperty("androidId", log.getCreatedOn());
        object.addProperty("createdOn", log.getCreatedOn());
        object.addProperty("androidId", NetworkUtils.getMacAddr());
        object.addProperty("deviceName", NetworkUtils.getDeviceName());
        object.addProperty("customDeviceName", NetworkUtils.getCustomDeviceName(context));
        object.addProperty("parentCode", log.getParentCode());
        return object;
    }

    public void setError(Throwable e) {

        this.error += "--------- Stack trace ---------\n\n";
        appendReport(e);
        this.error += "--------- Cause ---------\n\n";
        Throwable cause = e.getCause();
        appendReport(cause);
    }

    private void appendReport(Throwable cause) {
        if (cause != null) {
            this.error += cause.toString() + "\n\n";
            StackTraceElement[] arr = cause.getStackTrace();
            for (int i = 0; i < arr.length; i++) {
                this.error += "    " + arr[i].toString() + "\n";
            }
        }
        this.error += "-------------------------------\n\n";
    }

    public String getPage() {
        return page;
    }

    public void setPage(String page) {
        this.page = page;
    }
}

File: app/src/main/java/org/ole/planet/myplanet/model/RealmApkLog.kt
package org.ole.planet.myplanet.model

import android.content.Context
import com.google.gson.JsonObject
import io.realm.RealmObject
import io.realm.annotations.Ignore
import io.realm.annotations.PrimaryKey
import org.ole.planet.myplanet.utils.NetworkUtils

open class RealmApkLog : RealmObject() {
    @PrimaryKey
    var id: String? = null
    var userId: String? = null
    var type: String? = null
    var _rev: String? = null
    var error: String? = null
    var page: String? = null
    var parentCode: String? = null
    var version: String? = null
    var createdOn: String? = null
    var time: String? = null

    fun setError(e: Throwable) {
        val builder = StringBuilder(error ?: "")
        builder.append("--------- Stack trace ---------\n\n")
        appendReport(builder, e)
        builder.append("--------- Cause ---------\n\n")
        val cause = e.cause
        appendReport(builder, cause)
        error = builder.toString()
    }

    private fun appendReport(builder: StringBuilder, cause: Throwable?) {
        if (cause != null) {
            builder.append(cause.toString()).append("\n\n\n")
            val arr = cause.stackTrace
            for (i in arr.indices) {
                builder.append("    ").append(arr[i]).append("\n")
            }
        }
        builder.append("-------------------------------\n\n")
    }

    companion object {
        @Ignore
        const val ERROR_TYPE_CRASH = "crash"

        @JvmStatic
        fun serialize(log: RealmApkLog, context: Context): JsonObject {
            val `object` = JsonObject()
            `object`.addProperty("type", log.type)
            `object`.addProperty("error", log.error)
            `object`.addProperty("page", log.page)
            `object`.addProperty("time", log.time)
            `object`.addProperty("userId", log.userId)
            `object`.addProperty("version", log.version)
            `object`.addProperty("createdOn", log.createdOn)
            `object`.addProperty("androidId", NetworkUtils.getUniqueIdentifier())
            `object`.addProperty("deviceName", NetworkUtils.getDeviceName())
            `object`.addProperty("customDeviceName", NetworkUtils.getCustomDeviceName(context))
            `object`.addProperty("parentCode", log.parentCode)
            return `object`
        }
    }
}
